### PR TITLE
blockchain: improve StatePrefetcher by prefetching txs when fetcher works

### PR DESCRIPTION
**Pull request by ehnuje**
## Proposed changes

- Previous implementation of StatePrefetcher(=https://github.com/klaytn/klaytn/pull/832) won't be helpful when fetcher works, since it will process only a block at a time, since it pre-executes the next block, taken from the list of blocks.
- With this change, if single block is given to `insertChain`, before processing the block, it will be sent to prefetchTxWorker, which executes transactions in the block in a parallel manner. This can pre-cache the data needed for processing the block.

## Test Results
- EN0 is the one with original StatePrefetcher and EN2 is the one with original StatePrefetcher + fetcher optimized StatePrefetcher, EN2 shows about 20-30% of less execution time and LevelDB get time.

Screen Shot 2021-01-15 at 9 10 51 AM
Screen Shot 2021-01-15 at 9 11 23 AM
Screen Shot 2021-01-15 at 9 13 47 AM
Screen Shot 2021-01-15 at 9 14 11 AM
Screen Shot 2021-01-15 at 9 14 28 AM

- EN2 does not seem to use more resource than EN0, it only has 10% more goroutines than EN0

Screen Shot 2021-01-15 at 9 15 45 AM
Screen Shot 2021-01-15 at 9 16 15 AM
Screen Shot 2021-01-15 at 9 16 22 AM

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: blockchain/blockchain.go [hunk 1]
```diff
@@ -194,6 +195,15 @@ type BlockChain struct {
 	// Warm up
 	lastCommittedBlock uint64
 	quitWarmUp         chan struct{}
+
+	prefetchTxCh chan prefetchTx
```
Comment (aidan-kwon) [inline blockchain/blockchain.go:199]: I think it needs to be closed when `Blockchain.Stop()` is called.

File: blockchain/blockchain.go [hunk 2]
```diff
@@ -292,6 +309,19 @@ func NewBlockChain(db database.DBManager, cacheConfig *CacheConfig, chainConfig
 	return bc, nil
 }
 
+// prefetchTxWorker receives a block and a transaction index, which it pre-executes
+// to retrieve and cache the data for the actual block processing.
+func (bc *BlockChain) prefetchTxWorker() {
+	for followup := range bc.prefetchTxCh {
```
Comment (aidan-kwon) [inline blockchain/blockchain.go:315]: Does it need to be registered in `Blockchain.wg` wait group?

File: blockchain/state_prefetcher.go [hunk 3]
```diff
@@ -66,6 +70,29 @@ func (p *statePrefetcher) Prefetch(block *types.Block, stateDB *state.StateDB, c
...
+// PrefetchTx processes the state changes according to the Klaytn rules by running
+// a single transaction message using the statedb, but any changes are discarded. The
+// only goal is to pre-cache transaction signatures and state trie nodes. It is used
+// when fetcher works, so it fetches only a block.
+func (p *statePrefetcher) PrefetchTx(block *types.Block, ti int, stateDB *state.StateDB, cfg vm.Config, interrupt *uint32) {
+	var (
+		header = block.Header()
+	)
+	tx := block.Transactions()[ti]
+	// Iterate over and process the individual transactions
```
Comment (aidan-kwon) [inline blockchain/state_prefetcher.go:82]:
```suggestion
```

**Comment by winnie-byun** [comment]
LGTM except for the previous comments :)

**Comment by ehnuje** [comment]
@winnie-byun @aidan-kwon @KimKyungup I've updated the PR, please take a look.

**Review by aidan-kwon — APPROVED** [review]
LGTM. Please fix the test failures.

File: blockchain/blockchain.go [hunk 4]
```diff
@@ -272,6 +283,11 @@ func NewBlockChain(db database.DBManager, cacheConfig *CacheConfig, chainConfig
 			}
 		}
 	}
+
+	for i := 1; i <= runtime.NumCPU()/2; i++ {
```
Comment (KimKyungup) [inline blockchain/blockchain.go:287]: How about comparing severl goroutine number? 
```suggestion
	for i := 1; i <= runtime.NumCPU(); i++ {
```

**Review by winnie-byun — APPROVED** [review]

**Comment by ehnuje** [comment]
@aidan-kwon @winnie-byun @KimKyungup Test fails due to the newly added wait group.. 😂  Do you see any suspicious part?

File: blockchain/blockchain.go [hunk 5]
```diff
@@ -272,6 +283,12 @@ func NewBlockChain(db database.DBManager, cacheConfig *CacheConfig, chainConfig
...
+		bc.wg.Add(1)
+		go bc.prefetchTxWorker(i)
+	}
+
 	// Take ownership of this particular state
 	go bc.update()
 	go bc.gcCachedNodeLoop()
```
Comment (aidan-kwon) [inline blockchain/blockchain.go:294]:
```suggestion
	bc.gcCachedNodeLoop()
```
@ehnuje This line raise a race condition between `bc.wg.Done()` and `bc.wg.Add(1)` in `bc.gcCachedNodeLoop()`

File: blockchain/blockchain.go (hunk 4 above)
Comment (ehnuje) [reply blockchain/blockchain.go:287]: Let's merge this for now and then compare performance with different number of workers.

Comment (KimKyungup) [reply blockchain/blockchain.go:287]: Ok. Please test with different number.

**Review by KimKyungup — APPROVED** [review]

**Review by winnie-byun — APPROVED** [review]

**Review by aidan-kwon — APPROVED** [review]

**Pull request merged**